src/Raven.Server/Documents/StreamsTempFile.cs

```diff
@@ -28,24 +32,32 @@ public Stream StartNewStream()
                 throw new NotSupportedException("The temp file was already moved to reading mode");
 
             _previousInstance?.Flush();
-            if (_environment.Options.Encryption.IsEnabled)
-            {
-                _previousInstance = new TempCryptoStream(_file);
-            }
-            else
-            {
-                _previousInstance = new InnerPartStream(_file, this);
-            }
+            _previousInstance = _environment.Options.Encryption.IsEnabled
+                ? new InnerStream(new TempCryptoStream(_file), this)
+                : new InnerStream(new InnerPartStream(_file), this);
+
             return _previousInstance;
         }
 
-        public void Reset()
-        {
-            _reading = false;
+        public long Generation;
 
-            const int _128mb = 128 * 1024 * 1024;
-            if (_file.Length > _128mb)
-                _file.SetLength(_128mb);
+        public IDisposable Scope()
+        {
+            return new DisposableAction(() =>
+            {
+                _previousInstance = null;
+
+                Generation++;
+                _reading = false;
+                _file.ResetLength();
+
+                const int _128mb = 128 * 1024 * 1024;
+                if (_file.InnerStream.Length > _128mb)
+                {
+                    _file.InnerStream.SetLength(_128mb);
+                    _file.SetLength(_128mb);
+                }
```

Comment:
I find this piece of code difficult to understand. Let's talk about that to see if we can simplify it somehow.